mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -50,6 +50,7 @@ def _at_exit_callback(self) -> None:
             self._stop_data_logging_thread_event.set()
             # Waits till logging queue is drained.
             self._batch_logging_thread.join()
+            self._batch_logging_worker_threadpool.shutdown(wait=False)
```

Comment:
Why are we passing `wait=False` here if we're using `join()` for the other thread on line 52?

Reply:
this is by intention - `self._batch_logging_thread.join()` makes sure that data captured by the async logging queue is drained but not accepting new data,  `self._batch_logging_worker_threadpool.shutdown(wait=False)` returns the program control to user while keeping finishing submitted tasks in the background. Python threadpool is a bit strange that `wait=False` does not mean jobs are cancelled, but just blocking vs non-blocking. Differently in `flush`, we want to make sure all data are logged before returning the control, so we use `wait=True` there.

Reply:
> self._batch_logging_worker_threadpool.shutdown(wait=False) returns the program control to user while keeping finishing submitted tasks in the background

I don't understand what the implication of this is when it's called in `at_exit`. If we return control to program at exit, won't Python exit before processing all of the requests?

> this is by intention - self._batch_logging_thread.join() makes sure that data captured by the async logging queue is drained but not accepting new data,

Then, just to be safe, can't we change this to `wait=True`? Shouldn't it effectively be a no-op?

Reply:
> won't Python exit before processing all of the requests?
I see the confusion point now. The answer is no, python threadpool will keep running in the background, but the main program exits.

> Shouldn't it effectively be a no-op
it's not a no-op. let's say we run the program on terminal, `wait=True` will not return the cursor to user until everything is logged, `wait=False` will return the cursor immediately (almost immediately, because draining the queue is pretty fast). 

But to your proposal of changing this to wait=True, I am down to that, since it might be worth waiting for a few more seconds (with the optimization the hanging won't be long) to get all data logged before exiting compared to having things finished in background, which could be interrupted due to killing runtime or so.

Reply:
pushed a new commit to flip it to True